### ISSUE TITLE
Add java.time.Duration to Java APIs

### ISSUE THREAD

**Issue by ktoso** [akka/akka#24646]
After many many discussions: 

> We discussed this and are willing to give it to try to add java duration.

This is the super ticket, which should list all places where we want to add it, and then small PRs to address each of those.

**Comment by hepin1989** [comment]
Would not be much :)

![image](https://user-images.githubusercontent.com/501740/36886449-6687e5c0-1e27-11e8-918f-95a4d7f6d800.png)

**Comment by ralphlaude** [comment]
Hey Guys,
I would want work on this issue as my good first issue.
My understanding of the issue is to add java.time.Duration to Java APIs or JavaDSL. At the end we should not have scala.concurrent.duration.FiniteDuration in the Java APIs.
I would make the change in the class appearing in the Method Pattern Declaration list above.

**Comment by ktoso** [comment]
Hi there, thanks!

> At the end we should not have scala.concurrent.duration.FiniteDuration in the Java APIs.

Not entirely, so the methods have to remain there, as Akka maintains binary compatibility between releases. We can however add overloads accepting the Java one.

**Comment by ralphlaude** [comment]
Hi,
I am there.
So I will do it that way to maintain binary compatibility.

**Comment by ralphlaude** [comment]
Hi @ktoso,

can I work on this issue?
Thanks

**Comment by ktoso** [comment]
Absolutely, thanks a lot :)

**Comment by ralphlaude** [comment]
ok.
If I have questions I will come back to you :)

**Comment by ralphlaude** [comment]
Hi guys,
I would like to put the overloading methods in the sub-classes and let the trait Scala-pure concerning the java.time.Duration. Is it ok to solve it like this?
for example the overloading methods should be in akka.actor.AbstractScheduler, this way the trait akka.actor.Scheduler stays Scala-pure.
The same apply for akka.actor.AbstractFSM and akka.actor.FSM.
Thanks for your help.

**Comment by patriknw** [comment]
Sounds like a good approach when there is already such a class and that is what is exposed to end users.

I think that works for FSM but not Scheduler.

**Comment by hepin1989** [comment]
@ralphlaude The PR have been merged,which have done that:)

**Comment by patriknw** [comment]
@hepin1989 Then that was not completely correct, as far as I can see. https://github.com/akka/akka/pull/24651/files#r173815977
Would you be able to follow up?

**Comment by hepin1989** [comment]
@ralphlaude Will follow this? otherwise I will come up to this:)

**Comment by ralphlaude** [comment]
@patriknw I will follow the approach to let the trait Scala-pure if it is possible.
@hepin1989 I do not make a pull request
@patriknw I will try to follow up this https://github.com/akka/akka/pull/24651/files#r173815977

When I am done with the akka-actor module i will make a PullRequest with the status WorkInProgress. This way my work could be control. Is it the right way?

**Comment by ralphlaude** [comment]
Hey guys,
I would like a review for the changes done in the akka-actor module. I would like to make a WIP Pull Request. Is that fine like this? After the review i will extend the tests and write the documentation.
Any help or recommendation would be appreciate.
Thanks

**Comment by patriknw** [comment]
That’s fine, thanks for asking.

**Comment by ralphlaude** [comment]
Hey guys,
here is my WIP Pull Request - https://github.com/akka/akka/pull/24735

**Comment by ktoso** [comment]
TODO: remember to make sure we point out the new ones in documentation as well

**Comment by patriknw** [comment]
in general, better to update documentation together with each PR

**Comment by ralphlaude** [comment]
@patriknw I will update the documentation in the next step. I was not sure if my changes are right.
Is my PR already merge? Do I still need to update the documentation. Thanks for your answer.

**Comment by ralphlaude** [comment]
I am a bit confused because of the two merged PR. Should I continue with the issue in order module? So what is the next step?

**Comment by patriknw** [comment]
I'm currently fixing all places in Typed javadsl

**Comment by patriknw** [comment]
Related: https://github.com/akka/akka/issues/24805
JavaDurationConverters should be for FiniteDuration

**Comment by ralphlaude** [comment]
Hi guys,
actually i am done with the java.time.Duration in the JavaDSL in akka-actor module. The java documentation and the tests were changed accordingly.
I think, this pull request should be merged first before i start the next module. Is there another option to that?

**Comment by patriknw** [comment]
@chbatey I guess this was accidentally closed, or have we changed everywhere?

**Comment by ralphlaude** [comment]
thanks a lot and understood :).
I have a pending pull request (https://github.com/akka/akka/pull/24735/) and i want to if everything  for the akka-actor module is fine. If it is fine, then i will go further and may be the pull request could be merged. It is really a big big issue.
I still don't know if i could rebase the master on my branch. I would say no, then it is reason to merge the pending pull request.

**Comment by hepin1989** [comment]
https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala

**Comment by hepin1989** [comment]
akka.actor.ActorContext#setReceiveTimeout

**Comment by ralphlaude** [comment]
How should i understand that?

**Comment by ralphlaude** [comment]
Hey guys, what is the next steps here?

**Comment by patriknw** [comment]
@ralphlaude Try to find more places, for example `setReceiveTimeout` in `akka.actor.ActorContext`.

**Comment by patriknw** [comment]
@ralphlaude Yes, I think we should try to find solutions for those also. For the `setReceiveTimeout` I suggest that we add it in `AbstractActor.ActorContext` and also add another method there for the cancellation:

```
  /**
   * Cancel the sending of receive timeout notifications.
   */
  def cancelReceiveTimeout(): Unit
```

**Comment by ralphlaude** [comment]
@patriknw now i understand. i will investigate and make propositions for adding the java.time.Duration support for scala.concurrent.duration.Duration.
My question still remains concerning the branch i am working on. Should i continue on it or not?
Thanks a lot for your answer

**Comment by patriknw** [comment]
Your pull request has been merged to the master branch, so you should create a new branch and PR based on akka/akka's master.

**Comment by ralphlaude** [comment]
ok, thanks

**Comment by ralphlaude** [comment]
@patriknw concerning your point - I think we should try to find solutions for those also. For the setReceiveTimeout I suggest that we add it in AbstractActor.ActorContext and also add another method there for the cancellation:
```
  /**
   * Cancel the sending of receive timeout notifications.
   */
  def cancelReceiveTimeout(): Unit
```
I don't really understand why we should do this. Could you please give me the context?
My solution is to add this in akka.actor.ActorContext:
```
 def setReceiveTimeout(timeout: java.time.Duration): Unit =
setReceiveTimeout(timeout.asScala)
```
and this in akka.actor.AbstractActor.ActorContext:
```
 def setReceiveTimeout(timeout: java.time.Duration): Unit
```

Thanks for your answer.

**Comment by patriknw** [comment]
To cancel/remove the receive timeout you can use `setReceiveTimeout(scala.concurrent.duration.Duration.Undefined)`, but that is for the Scala API. There is nothing corresponding to `Undefined` in java.time, so we have to add another method for cancellation.

Those two methods can be added in `AbstractActor.ActorContext`, since that is the Java API.

**Comment by patriknw** [comment]
Change signature in `ReceiveTimeout` to:
```
override def cancelReceiveTimeout(): Unit =
```

It will then override the `Abstract.ActorContext#cancelReceiveTimeout`, and you already got the implementation of it for free.

**Comment by ralphlaude** [comment]
Thanks

**Comment by ralphlaude** [comment]
@patriknw: here is my pull request (https://github.com/akka/akka/pull/24993). I changed also the documentation. Please take a look.
If everything is fine, i will start adding the support for java.time.Duration in the next module.

**Comment by ralphlaude** [comment]
@johanandren: Do it need something else here, except putting something in the MiMa excludes?

**Comment by ralphlaude** [comment]
I would like to know which module to do next? Or may be the issue is done?
Thanks for hints.

**Comment by ralphlaude** [comment]
Hi guys, i would like to know what the next step is for this ticket. Thanks

**Comment by johanandren** [comment]
It could be that we are done, we have one open ticket, can you see if you can find any other API except the java TestKit only providing the Scala duration type?

**Comment by ralphlaude** [comment]
I will check all others API except the java TestKit

**Comment by ralphlaude** [comment]
@johanandren i found some Scala Duration in some Java classes. I pick up some and you take a look. Here is the list:
```
1- jdocs.actor.InitializationDocTest.java
TestKit.shutdownActorSystem(system) can be used instead of Await.ready(system.terminate(), Duration.create("5 seconds"))
expectNoMesage(java.time.Duration.ofSeconds(1)) can be used instead of expectNoMsg(Duration.create(1, TimeUnit.SECONDS))

2- jdocs.actor.DependencyInjectionDocTest.java
TestKit.shutdownActorSystem(system) can be used instead of Await.ready(system.terminate(), Duration.create("5 seconds"))

3- jdocs.actor.fsm.FSMDocTest.java

4- jdocs.ddata.ShoppingCart.java

5- jdocs.ddata.DistributedDataDocTest. Java

6- jdocs.event.EventBusDocTest.Java
expectNoMesage(java.time.Duration.ofSeconds(1)) can be used instead of expectNoMsg(Duration.create(1, TimeUnit.SECONDS))

7- jdocs.future.FutureDocTest.java here i am not sure

8- jdocs.stream.ActorSubscriberDocTest.java
expectTerminated(java.time.Duration.ofSeconds(10), worker) can be used instead of  expectTerminated(Duration.create(10, TimeUnit.SECONDS), worker)

9- jdocs.stream.StreamTestKitDocTest.java, probe could use java.time.Duration

10- jdocs.pattern. SupervisedAsk.java here i am not sure

11- jdocs.pattern. SupervisedAskSpec.java here i am not sure

12- jdocs.routing.RouterDocTest.Java
getContext().actorOf(new ScatterGatherFirstCompletedGroup(paths, java.time.Duration.ofSeconds(10)).props(), "router20") can be used instead of
getContext().actorOf(new ScatterGatherFirstCompletedGroup(paths, FiniteDuration.create(10, TimeUnit.SECONDS)).props(), "router20")
```

**Comment by ralphlaude** [comment]
@johanandren i think i forgot something here. The java.time.duration support is needed here:
```
1- akka.actor.OneForOneStrategy.scala should provide java.time.Duration support

2- akka.actor.AllForOneStrategy.scala should provide java.time.Duration support
```

**Comment by ktoso** [comment]
Thanks for the list :-) Would you want to go over those and make sure they can all use the Java types? (and if API is missing, potentially add those, e.g. like in AllForOne etc, if possible without breaking bin compat)

**Comment by ralphlaude** [comment]
Thanks for the answer. I will take care of :).

**Comment by ralphlaude** [comment]
Hi guys,
I create a Pull Request hier is link (https://github.com/akka/akka/pull/25136).

I had java.time.Duration support succesfully for most of the classes in the pull request.
I had java.time.Duration support in this class for the method forMax (`akka.actor.FSM.State#forMax`) and i don't know if it is fine.
The test `akka.stream.scaladsl.UnfoldResourceSourceSpec.scala` the method `"use dedicated blocking-io-dispatcher by default"` is broken and i would like to know how to fix it.

Please take a look.

**Comment by ralphlaude** [comment]
Hi guys, the new pull request is here (https://github.com/akka/akka/pull/25206).
Please take a look.
Thanks.

**Comment by ralphlaude** [comment]
Hi guys,
i would like to know the state of this issue. What is already done or what should be next step?
Thanks a lot

**Comment by johanandren** [comment]
I think it may be completed when the outstanding PRs are merged. Please let us know if you can find more duration using APIs that we have missed.

**Comment by ralphlaude** [comment]
I will take a look, but i will be difficult because i don't the whole overview.

**Comment by ralphlaude** [comment]
Hi guys,
i have 2 proposals concerning the java.time.Duration support for JavaDSL:

- Add util-object for Await with ready and result methods:
```
package akka.util

import scala.concurrent.Awaitable
import java.time.Duration

object Await {

  def ready[T](awaitable: Awaitable[T], atMost: Duration): awaitable.type = {
    import akka.util.JavaDurationConverters._
    scala.concurrent.Await.ready(awaitable, atMost.asScala)
  }

  def result[T](awaitable: Awaitable[T], atMost: Duration): T = {
    import akka.util.JavaDurationConverters._
    scala.concurrent.Await.result(awaitable, atMost.asScala)
  }
}
```

- Add new overloading methods in ```akka.pattern.PatternsCS``` for methods with ```akka.util.Timeout```. For example:
```
def ask(actor: ActorRef, message: Any, duration: java.time.Duration): CompletionStage[AnyRef] =
    ask(actor, message, Timeout.create(duration))

def askWithReplyTo(actor: ActorRef, messageFactory: japi.function.Function[ActorRef, Any], duration: java.time.Duration): CompletionStage[AnyRef] =
    askWithReplyTo(actor, messageFactory, Timeout.create(duration))

def ask(selection: ActorSelection, message: Any, duration: java.time.Duration): CompletionStage[AnyRef] =
    ask(selection, message, Timeout.create(duration))
```

What do you think about?

**Comment by ralphlaude** [comment]
Hi,
i saw in the class akka.event.LoggingAdapterTest.LogJavaTestKit.java the method expectLog the akka.testkit.javadsl.Testkit#expectMsgPF is still using the scala.concucurrent.duration.Duration instead of java.time.Duration. I am really sure if it is already handled in another pullrequest .

**Comment by patriknw** [comment]
> Add util-object for Await

Not needed because `Await` is for Scala API. `Await` is mostly for Scala's Future so Java API should have `CompletionStage`, which is possible to await via `.toCompletableFuture.get(timeout)`. 

> Add new overloading methods in akka.pattern.PatternsCS

`akka.util.Timeout` already has `create(java.time.Duration` but I think it's a good idea to support `java.time.Duration` directly in those ask methods. Please create PR.

**Comment by ralphlaude** [comment]
Thanks for the answers.
I will create the methods for ```akka.pattern.PatternsCS```. But i can use this pull request (https://github.com/akka/akka/pull/25206), why do i need a new one?

**Comment by ralphlaude** [comment]
Hi guys,

i created a new pull request, here is the link (https://github.com/akka/akka/pull/25277)

**Comment by ralphlaude** [comment]
Hi, i added new changes to the pull request. Please take a look.

**Comment by ralphlaude** [comment]
Hey, i think everything is done here. Thanks for an answer

**Comment by ralphlaude** [comment]
i think everything is done here.

**Comment by raboof** [comment]
Awesome! Thanks for all your help here!
